include/aie/Dialect/ADF/ADF.td

```diff
@@ -0,0 +1,170 @@
+//===- ADF.td - ADF dialect operation definitions ----------*- tablegen -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2021 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+//
+// Defines the operations and types of the ADF dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef ADF_TD
+#define ADF_TD
+
+include "mlir/IR/OpBase.td"
+include "mlir/IR/OpAsmInterface.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/Pass/PassBase.td"
+
+def ADF_Dialect : Dialect {
+  let name = "ADF";
+  let summary = "Types and operations for the ADF dialect";
+  let description = [{
+    This is a mlir dialect designed to capture the Xilinx Adaptive Data Flow
+    (ADF) Graph semantics. The dialect is designed to perform C++ ADF graph in
+    mind; it uses operations like create graph input, create graph output, and
+    create kernel along with an new interface type to reflect the real
+    connections in the C++ ADF graph.
+  }];
+  let cppNamespace = "::xilinx::ADF";
+}
+
+//===----------------------------------------------------------------------===//
+// Base ADF operation definition.
+//===----------------------------------------------------------------------===//
+class ADF_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<ADF_Dialect, mnemonic, traits>;
+
+
+//===----------------------------------------------------------------------===//
+// ADF Types
+//===----------------------------------------------------------------------===//
+def ADF_InterfaceType :
+    DialectType<ADF_Dialect, CPred<"$_self.isa<InterfaceType>()">,
+                "ADF interface type">;
+
+
+//===----------------------------------------------------------------------===//
+// ADF Operations
+//===----------------------------------------------------------------------===//
+
+def CreateADFGraphOp : ADF_Op<"graph", [NoSideEffect, NoTerminator]> {
+  let summary = "create an ADF graph";
+  let description = [{
+    Create an ADF graph region which consists of ADF operations. The graph
+    operation also need to specify several string attributes for C++ ADF graph
+    generation, including "graph_name", "graph_input_name", and
+    "graph_output_name". For example, 
+
```

Comment:
because I try to do just one mlir graph traversal, and the c++ code generation will need to generate the class data member before visiting the ginp and gout ADF Op. So the only way is to define it as the adf.graph_op attributes.